### PR TITLE
Sils/langdefmig

### PR DESCRIPTION
**Pull request by sils**

### CODE REVIEW

File: coalib/bearlib/languages/Language.py [hunk 1]
```diff
@@ -250,6 +281,12 @@ def __init__(self, *versions):
         else:
             self.versions = tuple(sorted(versions))
 
+    def __getattr__(self, item):
+        if len(self.versions) > 1:
```
Comment (Adrianzatreanu) [inline coalib/bearlib/languages/Language.py]: what if you specify 0?

Comment (sils) [inline coalib/bearlib/languages/Language.py]: that's fine, you can only specify 0 if there are None and then you can't spec one. Which is good

File: coalib/bearlib/languages/Language.py [hunk 2]
```diff
@@ -333,6 +338,16 @@ def attributes(self):
         """
         return list(self._attributes.keys())
 
+    def get_default_version(self):
+        """
+        Retrieves the latest version the user would want to choose from the
+        given versions in self.
+
+        (At a later point this might also retrieve a default version
```
Comment (Adrianzatreanu) [inline coalib/bearlib/languages/Language.py]: shouldnt this be a comment or so?

File: coalib/bearlib/languages/LanguageDefinition.py [hunk 3]
```diff
@@ -36,10 +30,20 @@ def __init__(self, language: str, coalang_dir=None):
...
         If no language exists, you will get a ``FileNotFoundError``:
 
-        >>> LanguageDefinition("BULLSHIT!")  # +ELLIPSIS
+        >>> LanguageDefinition("BULLSHIT!")
         Traceback (most recent call last):
          ...
-        FileNotFoundError: ...
+        FileNotFoundError
+
+        Custom coalangs are not longer supported. You can simply register your
```
Comment (Adrianzatreanu) [inline coalib/bearlib/languages/LanguageDefinition.py]: not -> no

File: coalib/bearlib/languages/definitions/C.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+from coalib.bearlib.languages.Language import Language
+
+
+@Language
+class C:
```
Comment (Adrianzatreanu) [inline coalib/bearlib/languages/definitions/C.py]: shouldnt these inherit from an abstract class?

File: coalib/bearlib/languages/definitions/C.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    extensions = '.c', '.h'
+    comment_delimiter = '//'
+    multiline_comment_delimiters = {'/*': '*/'}
+    string_delimiters = {'"': '"'}
```
Comment (Adrianzatreanu) [inline coalib/bearlib/languages/definitions/C.py]: i think chars are with ``'``, dont you need those too? what are these string delimiters used for anyway?

File: coalib/bearlib/languages/definitions/CPP.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
+from coalib.bearlib.languages.Language import Language
+
+
+@Language
+class CPP:
+    aliases = 'C++', 'C Plus Plus', 'CPlusPlus'
```
Comment (Adrianzatreanu) [inline coalib/bearlib/languages/definitions/CPP.py]: i think CPP works too

File: coalib/bearlib/languages/definitions/Java.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
+from coalib.bearlib.languages.Language import Language
+
+
+@Language
+class Java:
+    extensions = '.java'
+    comment_delimiter = '//'
+    multiline_comment_delimiters = {'/*': '*/'}
+    string_delimiters = {'"': '"', "'": "'"}
```
Comment (Adrianzatreanu) [inline coalib/bearlib/languages/definitions/Java.py]: not sure if relevant, but ``'`` is for chars, not strings :D

Comment (Adrianzatreanu) [inline coalib/bearlib/languages/definitions/Java.py]: doesnt java have keywords? i mean: http://docs.oracle.com/javase/tutorial/java/nutsandbolts/_keywords.html

File: coalib/bearlib/languages/definitions/C.py (hunk 5 above)
Comment (sils) [inline coalib/bearlib/languages/definitions/C.py]: chars != strings

they're used to find strings in your source code for whatever reason a bear wantt'

File: coalib/bearlib/languages/definitions/CPP.py (hunk 6 above)
Comment (sils) [inline coalib/bearlib/languages/definitions/CPP.py]: the name (CPP) works by default

File: coalib/bearlib/languages/definitions/Java.py (hunk 7 above)
Comment (Adrianzatreanu) [inline coalib/bearlib/languages/definitions/Java.py]: as in from your comment earlier: 

chars != strings

they're used to find strings in your source code for whatever reason a bear wantt'

Comment (sils) [inline coalib/bearlib/languages/definitions/Java.py]: but we haven't had them listed yet and I'm not going to resarch now :P we can complete them later

Comment (Adrianzatreanu) [inline coalib/bearlib/languages/definitions/Java.py]: you cannot use ``'`` for strings in java, so it shouldnt be here

Comment (userzimmermann) [inline coalib/bearlib/languages/definitions/Java.py]: @Adrianzatreanu It doesn't matter :)

File: coalib/bearlib/languages/definitions/CPP.py (hunk 6 above)
Comment (userzimmermann) [inline coalib/bearlib/languages/definitions/CPP.py]: `'CXX'` is missing. I saw it there some hours ago :)

**Comment by userzimmermann** [comment]
ack https://github.com/coala/coala/pull/3063/commits/6960d2fe95d0ff44dd1ac71d4261a21e765a7c8f

**Comment by userzimmermann** [comment]
ack https://github.com/coala/coala/pull/3063/commits/b02883a7a4abb9e98d85d28fa21617def1e88268

**Comment by userzimmermann** [comment]
@rultor merge

**Pull request merged**